### PR TITLE
Add MiniCPM-1B model support

### PR DESCRIPTION
**Pull request by Belugaaaa**
#### Before submitting

- [x] Lint code. If there are lint issues, please format the code first.

Validation passed locally:

- `pre-commit run --files ...` (including Black, isort, flake8, conflict checks, copyright checks, and dead-link checks)
- `bash -n scripts/minicpm/*.sh tests/integration_test/minicpm_sft_single_card.sh tests/integration_test/preprocess.sh`
- `python3 -m py_compile scripts/minicpm/*.py paddleformers/transformers/minicpm/*.py tests/transformers/minicpm/*.py`

- [x] Add test cases into `tests` folder. If there are codecov issues, please add tests cases first.

Added test coverage:

- `tests/transformers/minicpm/test_modeling.py`
- `tests/config/ci/minicpm_sft_single.yaml`
- `tests/integration_test/minicpm_sft_single_card.sh`
- H20 single-card workflow integration in `.github/workflows/fleet-model-test.yml`

### PR types

New features

### PR changes

Models, Docs

### Description

This PR adds PaddleFormers support for `openbmb/MiniCPM-1B-sft-bf16`, including:

- MiniCPM-1B config and model implementation.
- AutoConfig and AutoModelForCausalLM registration.
- HF-to-Paddle native checkpoint conversion and forward/generation alignment helpers.
- GSM8K SFT and ms-swift baseline scripts.
- Inference compiler benchmark helper.
- Deterministic tiny Paddle checkpoint/tokenizer generation for CI.
- Unit tests, including two-step KV-cache alignment coverage.
- A 10-step single-card SFT smoke test wired into the H20 single-card workflow.
- Acceptance status documentation.

The branch is rebased onto the latest `develop`, where MiniCPM4.1 is separated under `minicpm4_1`. This PR keeps the `minicpm` implementation specific to `openbmb/MiniCPM-1B-sft-bf16`.

Fleet/PP classes are intentionally not exposed in this PR. The generic GPT Fleet path does not implement MiniCPM's `scale_emb`, scaled residual connections, and pre-LM-head scaling. These paths should only be enabled after a MiniCPM-specific provider and ordinary-vs-Fleet/PP forward alignment test are added.

Validation results:

- Full MiniCPM-1B FP32 logits alignment:
  - `max_diff=4.57763671875e-05`
  - `mean_diff=8.224584234994836e-06`
  - `last_max_diff=3.0517578125e-05`
  - `last_mean_diff=7.700375135755166e-06`
- Generation alignment:
  - `first_10_tokens_match=True`
  - HF and Paddle first 10 generated token ids: `[11225, 72, 5, 2219, 8107, 1379, 8360, 1410, 11225, 72]`
- GSM8K 300-step PaddleFormers SFT completed:
  - Eval losses: step 50 `0.5329`, step 100 `0.5421`, step 150 `0.5408`, step 200 `0.5424`, step 250 `0.5483`, step 300 `0.5435`
  - Final train metrics: `train_loss=0.1506`, `train_steps_per_second=0.151`
- GSM8K 300-step ms-swift baseline completed:
  - Eval losses: step 50 `0.58973086`, step 100 `0.56501245`, step 150 `0.55724013`, step 200 `0.55477822`, step 250 `0.55431318`, step 300 `0.55419004`
  - Step-300 eval loss delta is about `0.0107`
- Unit tests passed locally: `Ran 34 tests ... OK (skipped=5)`.
- Complete 10-step tiny BF16 SFT smoke passed: `global_step=10`, `train_loss=3.9330`, and checkpoint saving completed.
- Inference compiler benchmark with `full_graph=True` and fused RMSNorm disabled for both paths:
  - Full-depth/full-width MiniCPM-1B architecture, seq len 128: `572.16%` speedup, `max_diff=0.00220`.
  - Full-depth/full-width MiniCPM-1B architecture, seq len 512: `109.96%` speedup, `max_diff=0.00250`.
  - These performance-only runs use a full-size random native checkpoint with the same model structure and vocabulary. Precision, generation, and training validations use the converted official checkpoint.

Known remaining acceptance item:

- Training-side compiler-on/off performance should be measured in the official acceptance environment. Inference-side compiler speedup exceeds 20% on the full MiniCPM-1B architecture.

### CODE REVIEW

**Comment by a31413510** [comment]
需要解决代码冲突

**Comment by Belugaaaa** [comment]
已同步最新 develop 并解决冲突。注意到 develop 中已经包含 MiniCPM 模型实现，因此本次冲突处理保留了官方 MiniCPM 实现，当前 PR 主要保留 MiniCPM 验收文档、SFT/ms-swift/benchmark 辅助脚本和 CI smoke 配置。

File: paddleformers/transformers/__init__.py [hunk 1]
```diff
@@ -193,6 +193,14 @@
...
     "llama.tokenizer": ["LlamaTokenizer", "Llama3Tokenizer"],
     "llama.tokenizer_fast": ["LlamaTokenizerFast"],
+    "minicpm.configuration": ["MiniCPMConfig"],
+    "minicpm.modeling": [
+        "MiniCPMDecoderLayer",
+        "MiniCPMModel",
+        "MiniCPMForCausalLM",
+        "MiniCPMForCausalLMPipe",
+        "MiniCPMForSequenceClassification",
+    ],
```
Comment (risemeup1111) [inline paddleformers/transformers/__init__.py]: P2 **优先级：P2**
**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

这里新增了一组 `minicpm.configuration` / `minicpm.modeling` key，但同一个 `import_structure` 后面已经有同名 key。Python dict 会保留后面的值，导致这组新增最终不会生效，特别是 `MiniCPMDecoderLayer` 不会通过 `from paddleformers.transformers import MiniCPMDecoderLayer` 暴露出来。

请在两种处理里选一种：如果希望顶层导出 `MiniCPMDecoderLayer`，把它补到后面已有的 `minicpm.modeling` block 并删掉这里这组重复 key；如果不希望顶层导出它，则直接删掉这里这组重复 key，避免留下死配置。有效的 block 应类似：

```python
"minicpm.modeling": [
    "MiniCPMDecoderLayer",
    "MiniCPMModel",
    "MiniCPMForCausalLM",
    "MiniCPMForCausalLMPipe",
    "MiniCPMForSequenceClassification",
],
```

**Review by risemeup1111 — COMMENTED** [review]
已复查，当前没有需要阻塞合入的代码问题；详细意见见 inline review comments。

- P3 **优先级：P3** 非行级：PR 描述里仍保留了“新增 MiniCPM config/model implementation / 单测”的旧表述，但当前 head 的 diff 重点已经变成验收文档、辅助脚本和 CI smoke 配置。建议把描述收敛到当前实际内容，并明确说明 MiniCPM 模型实现已在 develop 中。**处理要求：请更新后回复确认。**

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by Belugaaaa** [comment]
已完成。修复了 MiniCPM 重复的 import_structure 代码块，并在同步 develop 分支后更新 PR 描述，使其与当前改动范围保持一致。

**Review by risemeup1111 — APPROVED** [review]
已复查，这次编辑主要是把 PR 描述和 MiniCPM 导出结构进一步对齐；上一轮关于 `import_structure` 的问题已修正，当前未发现新的问题。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by a31413510** [comment]
提交内容中没有模型实现，是因为什么？

**Comment by Belugaaaa** [comment]
您好，更新一下当前状态：之前同步 develop 解决冲突时，我误以为 #4421 中的 `minicpm` 实现与本 PR 的 MiniCPM-1B 是同一模型，因此临时保留了 develop 中的实现，导致当时的 PR diff 中模型实现被移除。

现已确认 #4421 对应 MiniCPM4.1，与本 PR 目标 `openbmb/MiniCPM-1B-sft-bf16` 不是同一具体模型。当前 PR 分支已恢复到原始 MiniCPM-1B 实现版本。
后续等待 MiniCPM4.1 在 develop 中正确改名/拆分后，再 rebase 解决路径冲突。

File: paddleformers/transformers/minicpm/modeling.py [hunk 2]
```diff
@@ -0,0 +1,1086 @@
...
+    SequenceClassifierOutputWithPast,
+    TokenClassifierOutput,
+)
+from ..model_utils import PretrainedModel, register_base_model
+from ..modeling_rope_utils import ROPE_INIT_FUNCTIONS, dynamic_rope_update
+from .configuration import MiniCPMConfig
+
+
+@dataclass
+class MiniCPMModelProvider(GPTModelProvider):
```
Comment (risemeup1111) [inline paddleformers/transformers/minicpm/modeling.py]: P1 这里的 provider 被 `MiniCPMForCausalLMFleet` 和 `MiniCPMForCausalLMPipe` 直接交给通用 GPT builder，但 MiniCPM 普通路径分别在 embedding、两处 residual 和 LM head 应用了 `scale_emb`、`scale_depth / sqrt(num_hidden_layers)`、`hidden_size / dim_model_base` 三组缩放。当前固定的 PaddleFleet 版本中，通用 `GPTEmbedding`、BDA 和 `GPTLMHead` 分别只做原始 embedding、`residual + branch` 和直接投影，也不会读取这三个配置项；因此同一份权重经 Fleet/PP 路径会得到不同 logits 和训练更新。请提供 MiniCPM 专用 layer spec/provider 补齐三处计算，并增加普通路径与 Fleet/PP 的前向对齐测试；否则先不要暴露这两个路径。

File: paddleformers/transformers/minicpm/modeling.py [hunk 3]
```diff
@@ -0,0 +1,1086 @@
...
+        if inputs_embeds is None:
+            # [bs, seq_len, dim]
+            inputs_embeds = self.embed_tokens(input_ids).astype(self.embed_tokens.weight.dtype) * self.config.scale_emb
+
+        if use_cache and past_key_values is None:
+            past_key_values = DynamicCache(config=self.config)
+        cache_length = past_key_values.get_seq_length() if past_key_values is not None else 0
+
+        if position_ids is None:
+            position_ids = paddle.arange(seq_length, dtype="int64").expand((batch_size, seq_length))
```
Comment (risemeup1111) [inline paddleformers/transformers/minicpm/modeling.py]: P2 上面已经算出了 `cache_length`，但这里默认位置仍固定从 0 开始。调用方按可选参数契约复用第一次返回的 `past_key_values`、第二次只传一个新 token 且省略 `position_ids` 时，这个 token 会再次使用位置 0，而不是缓存末尾的位置，导致增量 logits 错误。请从 `cache_length` 生成区间，并补一个“两步缓存结果与整段前向末 token 对齐”的测试。

File: tests/integration_test/minicpm_sft_single_card.sh [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+export FLAGS_cudnn_deterministic=1
+export FLAGS_use_stride_compute_kernel=False
+export CUDA_VISIBLE_DEVICES=0
+
+unset http_proxy https_proxy
+
+log_file=minicpm_sft_single_card.txt
+
+set +e
+coverage run "$(which paddleformers-cli)" train "$config_yaml" 2>&1 | tee ./"${log_file}"
```
Comment (risemeup1111) [inline tests/integration_test/minicpm_sft_single_card.sh]: P2 这个 smoke 脚本目前只在自身出现，现有 workflow 和测试入口都没有调用它，因此 PR 检查不会执行这里的 10 步训练，也无法发现训练回归。请把它接入对应的单卡 CI job，并同时准备脚本依赖的 `${CACHE_DIR}/minicpm/tiny-random-minicpm` 资产，确保该 smoke 真正由检查运行。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
|---|---|---|---|
| 1 | [modeling.py:64](https://github.com/PaddlePaddle/PaddleFormers/pull/4601#discussion_r3711330905) | P1 | 🚧 |
| 2 | [modeling.py:601](https://github.com/PaddlePaddle/PaddleFormers/pull/4601#discussion_r3711332760) | P2 | 🚧 |
| 3 | [minicpm_sft_single_card.sh:43](https://github.com/PaddlePaddle/PaddleFormers/pull/4601#discussion_r3711334625) | P2 | 🚧 |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by a31413510** [comment]
需要解决代码冲突问题

**Comment by a31413510** [comment]
需要解决代码冲突

File: paddleformers/transformers/minicpm/modeling.py (hunk 3 above)
Comment (Belugaaaa) [reply paddleformers/transformers/minicpm/modeling.py]: Done. 默认 `position_ids` 现从 `cache_length` 开始生成，并新增 `test_cached_forward_matches_full_forward`，验证“前缀缓存 + 单 token”与整段前向最后 token 的 logits 对齐；本地完整单测 `Ran 34 tests ... OK (skipped=5)`。

File: tests/integration_test/minicpm_sft_single_card.sh (hunk 4 above)
Comment (Belugaaaa) [reply tests/integration_test/minicpm_sft_single_card.sh]: Done. 已将 MiniCPM-1B 10 步 SFT smoke 接入 `.github/workflows/fleet-model-test.yml` 的 H20 单卡 job。`tests/integration_test/preprocess.sh` 会确定性生成 `${CACHE_DIR}/minicpm/tiny-random-minicpm`（含 Paddle 权重和最小 tokenizer），不再依赖待上传的外部 tiny 资产；smoke 对训练失败会直接返回非零状态。

File: paddleformers/transformers/minicpm/modeling.py (hunk 2 above)
Comment (Belugaaaa) [reply paddleformers/transformers/minicpm/modeling.py]: Done. 已移除 `MiniCPMModelProvider`、`MiniCPMForCausalLMFleet`、`MiniCPMForCausalLMPipe` 及相关顶层导出，当前仅保留已完成精度验证的普通 MiniCPM-1B 路径。后续仅在补齐 MiniCPM 专用 embedding/residual/LM-head scaling provider 及普通路径与 Fleet/PP 前向对齐测试后再开放该路径。

**Comment by Belugaaaa** [comment]
已同步并 rebase 到最新 `develop`（`93ffb5e8`），当前 PR 已无代码冲突。三项 inline review 已逐条处理并回复：移除未对齐的 Fleet/PP 暴露、修复 KV-cache 默认位置并新增对齐测试、将带自动 tiny 资产生成的 10 步 SFT smoke 接入 H20 单卡 CI。完整 MiniCPM 单测本地通过：`Ran 34 tests ... OK (skipped=5)`；pre-commit 全部通过。烦请复查，谢谢。

**Review by risemeup1111 — APPROVED** [review]

**Comment by Belugaaaa** [comment]
已继续补充验证并推送 commit `f88ba9a1`：修复了 CI smoke 中 `yq strenv` 变量未导出导致路径为空的问题，完整 10-step BF16 SFT smoke 已通过（`global_step=10`, `train_loss=3.9330`）；同时增强编译 benchmark 的低峰值加载和长序列精度比较。当前 `full_graph=True` 且动态/静态均关闭 fused RMSNorm 时，全深全宽 MiniCPM-1B 同结构性能 checkpoint 在 seq 128/512 的 GPU 推理加速分别为 `572.16%`/`109.96%`，`max_diff` 分别为 `0.00220`/`0.00250`。随机 checkpoint 仅用于同结构性能验证，模型精度、生成和训练结论仍使用官方权重。PR 描述及验收文档已同步更新。
